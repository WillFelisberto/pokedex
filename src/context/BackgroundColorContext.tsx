'use client';

import { createContext, useContext, useState } from 'react';

const BackgroundColorContext = createContext({
  backgroundColor: '#ffffff',
  setBackgroundColor: (_color: string) => {}
});

export function BackgroundColorProvider({ children }: { children: React.ReactNode }) {
  const [backgroundColor, setBackgroundColor] = useState('#ffffff');

  return (
    <BackgroundColorContext.Provider value={{ backgroundColor, setBackgroundColor }}>
      <div style={{ backgroundColor, minHeight: '100vh', alignContent: 'center' }}>{children}</div>
    </BackgroundColorContext.Provider>
  );
}

export function useBackgroundColor() {
  return useContext(BackgroundColorContext);
}
