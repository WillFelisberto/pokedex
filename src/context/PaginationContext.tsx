'use client';
import { createContext, useContext, useState, ReactNode } from 'react';

interface PaginationContextType {
  currentPage: number;
  setCurrentPage: (page: number) => void;
}

const PaginationContext = createContext<PaginationContextType | null>(null);

interface PaginationProviderProps {
  children: ReactNode;
}

export const PaginationProvider: React.FC<PaginationProviderProps> = ({ children }) => {
  const [currentPage, setCurrentPage] = useState<number>(1);

  return (
    <PaginationContext.Provider value={{ currentPage, setCurrentPage }}>
      {children}
    </PaginationContext.Provider>
  );
};

export const usePagination = (): PaginationContextType => {
  const context = useContext(PaginationContext);
  if (!context) {
    throw new Error('usePagination must be used within a PaginationProvider');
  }

  return context;
};
