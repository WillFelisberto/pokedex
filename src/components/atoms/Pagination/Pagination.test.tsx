import { PaginationProvider } from '@/context/PaginationContext';
import { render, screen, fireEvent } from '@/tests/test-utils';

import { Pagination } from '.';

const renderWithProvider = (ui: React.ReactNode) => {
  return render(<PaginationProvider>{ui}</PaginationProvider>);
};

describe('<Pagination />', () => {
  const mockGoToPage = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render pagination with correct number of pages', () => {
    const { container } = renderWithProvider(
      <Pagination currentPage={3} totalPages={10} goToPage={mockGoToPage} isFetching={false} />
    );
    expect(screen.getByTestId('pagination-button-3')).toHaveClass('font-bold');
    expect(screen.getByTestId('pagination-button-2')).toBeInTheDocument();
    expect(screen.getByTestId('pagination-button-4')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should call goToPage when clicking on a page number', () => {
    renderWithProvider(
      <Pagination currentPage={3} totalPages={10} goToPage={mockGoToPage} isFetching={false} />
    );

    fireEvent.click(screen.getByText('4'));
    expect(mockGoToPage).toHaveBeenCalledWith(4);
  });

  it('should disable previous button on first page', () => {
    renderWithProvider(
      <Pagination currentPage={1} totalPages={10} goToPage={mockGoToPage} isFetching={false} />
    );

    expect(screen.getByTestId('previous-button')).toBeDisabled();
  });

  it('should disable next button on last page', () => {
    renderWithProvider(
      <Pagination currentPage={10} totalPages={10} goToPage={mockGoToPage} isFetching={false} />
    );

    expect(screen.getByTestId('next-button')).toBeDisabled();
  });

  it('should disable   buttons while fetching', () => {
    renderWithProvider(
      <Pagination currentPage={3} totalPages={10} goToPage={mockGoToPage} isFetching={true} />
    );

    expect(screen.getByTestId('previous-button')).toBeDisabled();
    expect(screen.getByTestId('next-button')).toBeDisabled();
  });
});
