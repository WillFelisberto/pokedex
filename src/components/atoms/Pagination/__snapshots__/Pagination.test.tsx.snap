// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render pagination with correct number of pages 1`] = `
<div>
  <div
    class="relative flex items-center py-4 w-full max-w-full overflow-hidden flex-row justify-center lg:flex-col lg:h-screen md:h-auto"
  >
    <button
      class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-2 py-2 bg-transparent text-white rounded-md hover:scale-110 disabled:opacity-50 hover:text-red-500 transition"
      data-testid="previous-button"
    >
      <span>
        ChevronLeft Icon
        ChevronLeft Icon
      </span>
    </button>
    <div
      class="flex space-x-1 md:flex-row lg:flex-col md:space-x-2 lg:space-x-0 md:space-y-0 lg:space-y-2 overflow-hidden"
    >
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-1"
      >
        <span>
          1
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-2"
      >
        <span>
          2
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent !scale-125 font-bold"
        data-testid="pagination-button-3"
      >
        <span>
          3
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-4"
      >
        <span>
          4
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-5"
      >
        <span>
          5
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-6"
      >
        <span>
          6
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-7"
      >
        <span>
          7
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-8"
      >
        <span>
          8
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-9"
      >
        <span>
          9
        </span>
      </button>
      <button
        class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-4 py-2 rounded-md transition bg-transparent text-white hover:scale-125 disabled:opacity-50"
        data-testid="pagination-button-10"
      >
        <span>
          10
        </span>
      </button>
    </div>
    <button
      class="flex items-center justify-center gap-2 px-4 py-2 rounded-lg bg-blue-600 text-white transition disabled:opacity-50 disabled:cursor-not-allowed px-2 py-2 bg-transparent text-white rounded-md disabled:opacity-50 hover:text-red-500 transition"
      data-testid="next-button"
    >
      <span>
        ChevronDown Icon
        ChevronLeft Icon
      </span>
    </button>
  </div>
</div>
`;
